[Merged by Bors] - feat(Order.LiminfLimsup): add limsup_max

D-Thomine opened the pull request:
Add characterization of limsup and limsup of max.

---

This PR adds to Order.LiminfLimsup:

- Characterizations of the limsup and liminf in Conditionally complete linear orders: `limsup_le_iff`, `le_limsup_iff`, `le_liminf_iff`, `liminf_le_iff`.
- Identities such as limsup max = max limsup: `limsup_max`, `liminf_min`. Some lemmas about bounded/cobounded functions to handle limsup max.
- Same identities with supremum/infimum indexed over a Finset: `limsup_finset_sup`, `liminf_finset_inf`, `limsup_finset_sup'`, `liminf_finset_inf'`.
- ~Some identities defined on Conditionally complete linear orders now have versions on Complete linear orders. The autocomplete tactic doesn't work very well, so in practice they can be very helpful. Example: `limsup_le_limsup'`.~

This file is in dire need of a refactor. It is getting too long ; the linter fails because the file has reached almost 1800 lines (I am not sure if it makes sense to disable the flag as long as we don't go over the limit of 1900 lines). There are many less immediate issues which would warrant some extensive work.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks, and is labeled with `awaiting-review`.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

pitmonticone commented:
Hi @D-Thomine, updated `style-exceptions.txt` as discussed [here](https://github.com/mseri/BET/issues/55#issuecomment-2217706009) and [here](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Porting.20TopologicalEntropy.20into.20Mathlib/near/450168092).

D-Thomine commented on Mathlib/Order/LiminfLimsup.lean:
```diff
@@ -617,6 +617,10 @@ theorem limsup_le_limsup {α : Type*} [ConditionallyCompleteLattice β] {f : Fil
   limsSup_le_limsSup hu hv fun _ => h.trans
 #align filter.limsup_le_limsup Filter.limsup_le_limsup
 
+theorem limsup_le_limsup' {α} [CompleteLattice β] {f : Filter α} {u v : α → β} (h : u ≤ᶠ[f] v) :
```
My feeling is that autoparam works well with e.g. the `exact` tactic, but not the `apply` tactic. But, yes, I agree that the right solution is to fix autoparam instead of adding duplicate lemmas (and then delete the duplicate lemmas).

I have a dozen of examples of autoparam not firing. I will try to write some minimal examples for Zulip.

D-Thomine replied:
For my issues with the automatic fill-in: I've opened a [Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/239415-metaprogramming-.2F-tactics/topic/isBoundedDefault.20not.20triggering/near/450437091).

In the meantime, these specialized versions are removed. Better to work now with a wonky tool which will, hopefully, be fixed, than to have to delete lemmas (with all the downstream effects) later.

D-Thomine commented:
Opening this PR for review. Apart from some golfing:

- @pitmonticone has added a style exception so that the linter stops complaining about the length of the file (thanks!).
- I've removed the duplicate lemmas which gave simpler versions of the theorem on Complete lattices / Complete linear orders.
- I've modified some lemmas about IsBoundedUnder/IsCoboundedUnder for finite sup/inf. One lemmas about the sup was not useful, while I've added two lemmas about the inf which are.

sgouezel commented on Mathlib/Order/LiminfLimsup.lean:
```diff
@@ -1533,3 +1579,161 @@ theorem OrderIso.liminf_apply {γ} [ConditionallyCompleteLattice β] [Conditiona
+  · refine (limsup_le_iff cobddmax bddmax).2 (fun b hb ↦ ?_)
+    have hu := eventually_lt_of_limsup_lt (lt_of_le_of_lt (le_max_left _ _) hb) h₃
+    have hv := eventually_lt_of_limsup_lt (lt_of_le_of_lt (le_max_right _ _) hb) h₄
+    refine mem_of_superset (inter_mem hu hv) (fun _ ↦ ?_)
+    simp
```
```suggestion
    exact mem_of_superset (inter_mem hu hv) (fun _ ↦ by simp)
```

D-Thomine replied:
Done.

sgouezel commented on Mathlib/Order/LiminfLimsup.lean:
```diff
@@ -1533,3 +1579,161 @@ theorem OrderIso.liminf_apply {γ} [ConditionallyCompleteLattice β] [Conditiona
+  simp only [Finset.sup'_apply, sup'_le_iff]
+  exact fun i i_s ↦ le_trans (h i i_s) (le_sup' m i_s)
+
+theorem isCoboundedUnder_le_finset_sup' [LinearOrder β] {f : Filter α} {F : ι → α → β}
```
For symmetry, what about a `sup` version for this one also?

D-Thomine replied:
It is already in the file, under the name `isCobounded_le_of_bot` :)

sgouezel commented on Mathlib/Order/LiminfLimsup.lean:
```diff
@@ -1533,3 +1579,161 @@ theorem OrderIso.liminf_apply {γ} [ConditionallyCompleteLattice β] [Conditiona
+    (h₁ : ∀ i ∈ s, f.IsCoboundedUnder (· ≤ ·) (F i))
+    (h₂ : ∀ i ∈ s, f.IsBoundedUnder (· ≤ ·) (F i)) :
+    limsup (fun a ↦ sup s F a) f = sup s (fun i ↦ limsup (F i) f) := by
+  rcases eq_or_neBot f with (rfl | _)
```
instead of doing twice the same proof, is it possible to deduce this one from the `sup'` version, separating the cases where `s` is empty or not?

D-Thomine replied:
Nice idea, it works very well (and made me notice an imprecision in some statements : I now write `limsup (fun a ↦ sup s (fun i ↦ F i a)) f` instead of `limsup (fun a ↦ sup s F a) f`).

sgouezel reviewed (commented):
Looks good, thanks!

sgouezel commented on Mathlib/Order/LiminfLimsup.lean:
```diff
@@ -1533,3 +1579,144 @@ theorem OrderIso.liminf_apply {γ} [ConditionallyCompleteLattice β] [Conditiona
+
+theorem limsup_finset_sup' [ConditionallyCompleteLinearOrder β] {f : Filter α}
+    {F : ι → α → β} {s : Finset ι} (hs : s.Nonempty)
+    (h₁ : ∀ i ∈ s, f.IsCoboundedUnder (· ≤ ·) (F i))
+    (h₂ : ∀ i ∈ s, f.IsBoundedUnder (· ≤ ·) (F i)) :
```
```suggestion
    (h₁ : ∀ i ∈ s, f.IsCoboundedUnder (· ≤ ·) (F i) := by exact fun i hi ↦ by isBoundedDefault)
    (h₂ : ∀ i ∈ s, f.IsBoundedUnder (· ≤ ·) (F i) := by exact fun i hi ↦ by isBoundedDefault) :
```
Same thing below.

D-Thomine replied:
Nice idea. Tested, and in works!

sgouezel commented on Mathlib/Order/LiminfLimsup.lean:
```diff
@@ -1533,3 +1579,144 @@ theorem OrderIso.liminf_apply {γ} [ConditionallyCompleteLattice β] [Conditiona
+    simp only [Finset.sup'_apply, le_sup'_iff]
+    use i, i_s
+
+theorem limsup_finset_sup [ConditionallyCompleteLinearOrderBot β] {f : Filter α}
```
Is it possible instead to state this assuming `[ConditionallyCompleteLinearOrder β] [OrderBot β]` instead? In this way, it would be also dualizable to a liminf version, which is not possible currently.

D-Thomine replied:
This requires a tweak to another file. The lemma `csSup_univ` required a `ConditionallyCompleteLinearOrderBot`, but its assumptions can be weakened to `ConditionallyCompleteLinearOrder` + `OrderBot`. With this modification, which should not break anything else, this works.

sgouezel commented:
bors r+
Thanks!
